cmd/generator/main.go

```diff
@@ -0,0 +1,219 @@
+			"mattermost-plugin-custom-attributes",
+			"mattermost-plugin-skype4business",
+			"mattermost-plugin-aws-SNS",
+			"mattermost-plugin-gitlab",
+			"mattermost-plugin-nps",
+		}
+
+		plugins := []*model.Plugin{}
+
+		for _, repositoryName := range repositoryNames {
+			logger.Debugf("querying repository %s", repositoryName)
+
+			plugin, err := getReleasePlugin(ctx, client, repositoryName, includePreRelease)
+			if err != nil {
+				return errors.Wrapf(err, "failed to release plugin for repository %s", repositoryName)
+			}
+
+			plugins = append(plugins, plugin)
+		}
+
+		encoder := json.NewEncoder(os.Stdout)
+		err := encoder.Encode(plugins)
+		if err != nil {
+			return errors.Wrap(err, "failed to encode plugins result")
+		}
+
+		return nil
+	},
+}
+
+func getReleasePlugin(ctx context.Context, client *github.Client, repositoryName string, includePreRelease bool) (*model.Plugin, error) {
+	logger := logger.WithField("repository", repositoryName)
+
+	repository, _, err := client.Repositories.Get(ctx, "mattermost", repositoryName)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get repository")
+	}
+
+	latestRelease, err := getLatestRelease(ctx, client, repositoryName, includePreRelease)
+	if err != nil {
+		return nil, err
+	}
+	if latestRelease == nil {
+		logger.Warnf("no latest release found for repository")
+		return nil, nil
+	}
+
+	var releaseName string
+	if latestRelease.GetName() == "" {
+		releaseName = latestRelease.GetTagName()
+	} else {
+		releaseName = fmt.Sprintf("%s (%s)", latestRelease.GetName(), latestRelease.GetTagName())
+	}
+	logger.Debugf("found latest release %s", releaseName)
+
+	downloadURL := ""
+	for _, releaseAsset := range latestRelease.Assets {
+		if strings.HasSuffix(releaseAsset.GetName(), ".tar.gz") {
+			downloadURL = releaseAsset.GetBrowserDownloadURL()
+		}
+	}
+
+	if downloadURL == "" {
+		logger.Warnf("Failed to find plugin asset release %s", releaseName)
+		return nil, nil
+	}
+	logger.Debugf("fetching download url %s", downloadURL)
+
+	resp, err := http.Get(downloadURL)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to download plugin bundle for release %s", releaseName)
+	}
+	defer resp.Body.Close()
+
+	plugin := model.Plugin{
+		HomepageURL:       repository.GetHTMLURL(),
+		DownloadURL:       downloadURL,
+		DownloadSignature: []byte{},
+	}
+
+	gzBundleReader, err := gzip.NewReader(resp.Body)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to read gzipped plugin bundle for release %s", releaseName)
+	}
+
+	bundleReader := tar.NewReader(gzBundleReader)
+	for {
+		hdr, err := bundleReader.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to read plugin bundle for release %s", releaseName)
+		}
+
+		if path.Base(hdr.Name) != "plugin.json" {
+			continue
+		}
+		manifest := mattermostModel.ManifestFromJson(bundleReader)
+
+		plugin.Manifest = manifest
+		break
+	}
+
+	if plugin.Manifest == nil {
```

Comment:
This is great that is validating the bundle before going to the marketplace. Will we be adding signature verification as well?

Reply:
That's a really good idea! No reason to distribute something that fails validation from the get go.